File: 01_Data_Structures/Backtracking/LC_079_Word_Search.java
/**
 * @file LC_079_Word_Search.java
 * @author BalaSeethaRamanjaneyulu
 * @brief Solution for LeetCode 079: Word Search
 * 
 * Problem: Find if a word exists in a grid of characters (adjacent horizontally/vertically).
 *          Each cell can be used only once per word.
 *
 * Platform: LeetCode #079
 * Approach: Backtracking (DFS on Grid)
 * Time:     O(M * N * 3^L)
 * Space:    O(L)
 * Link:     https://leetcode.com/problems/word-search/
 */

public class LC_079_Word_Search {

    /**
     * DFS exploring the grid for the target word.
     *
     * @param r      Row index.
     * @param c      Column index.
     * @param index  Character index in word.
     * @param board  Grid of characters.
     * @param word   Target word.
     * @return boolean True if found.
     */
    private boolean backtrack(int r, int c, int index, char[][] board, String word) {
        if (index == word.length()) return true;

        if (r < 0 || r >= board.length || c < 0 || c >= board[0].length || board[r][c] != word.charAt(index)) {
            return false;
        }

        char temp = board[r][c];
        board[r][c] = '#'; // Mark as visited

        boolean found = backtrack(r + 1, c, index + 1, board, word) ||
                        backtrack(r - 1, c, index + 1, board, word) ||
                        backtrack(r, c + 1, index + 1, board, word) ||
                        backtrack(r, c - 1, index + 1, board, word);

        board[r][c] = temp; // Backtrack: Restore state
        return found;
    }

    public boolean exist(char[][] board, String word) {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (backtrack(i, j, 0, board, word)) return true;
            }
        }
        return false;
    }

    // ─── Driver ──────────────────────────────────────────────────────────────
    public static void main(String[] args) {
        LC_079_Word_Search sol = new LC_079_Word_Search();

        char[][] board = {
            {'A', 'B', 'C', 'E'},
            {'S', 'F', 'C', 'S'},
            {'A', 'D', 'E', 'E'}
        };
        
        System.out.println("Test 1 — Result: " + sol.exist(board, "ABCCED")); // true
        System.out.println("Test 2 — Result: " + sol.exist(board, "SEE"));    // true
        System.out.println("Test 3 — Result: " + sol.exist(board, "ABCB"));   // false
    }
}

File: 01_Data_Structures/Backtracking/LC_079_Word_Search.cpp
/**
 * @file LC_079_Word_Search.cpp
 * @author BalaSeethaRamanjaneyulu
 * @brief Solution for LeetCode 079: Word Search
 *
 * Problem: Given an m x n grid of characters board and a string word, return 
 *          true if word exists in the grid. The word can be constructed from 
 *          letters of sequentially adjacent cells, where adjacent cells are 
 *          horizontally or vertically neighboring. The same letter cell may 
 *          not be used more than once.
 *
 * Platform: LeetCode #079
 * Approach: Backtracking (DFS on Grid)
 * Time:     O(M * N * 3^L) — where M, N are board dims, L is word length.
 *           At each step, we have 4 directions, but we don't go back, so 3 choices.
 * Space:    O(L) — recursion stack depth matches word length.
 * Link:     https://leetcode.com/problems/word-search/
 */

#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    /**
     * @brief Performs DFS starting from a specific cell to find the word.
     *
     * @param r      Current row.
     * @param c      Current column.
     * @param index  Current character index in the word.
     * @param board  The character grid.
     * @param word   The target string.
     * @return true  Found the word.
     * @return false Word not found.
     */
    bool backtrack(int r, int c, int index, vector<vector<char>>& board, string& word) {
        // Base case: all characters found
        if (index == word.length()) return true;

        // Boundary and mismatch check
        if (r < 0 || r >= board.size() || c < 0 || c >= board[0].size() || board[r][c] != word[index]) {
            return false;
        }

        // Choose: Mark as visited (using a temporary char to avoid extra space)
        char temp = board[r][c];
        board[r][c] = '#';

        // Explore: 4 directions (Up, Down, Left, Right)
        bool found = backtrack(r + 1, c, index + 1, board, word) ||
                     backtrack(r - 1, c, index + 1, board, word) ||
                     backtrack(r, c + 1, index + 1, board, word) ||
                     backtrack(r, c - 1, index + 1, board, word);

        // Un-choose (Backtrack): Restore character
        board[r][c] = temp;

        return found;
    }

    bool exist(vector<vector<char>>& board, string word) {
        int rows = board.size();
        int cols = board[0].size();

        // Start DFS from every cell
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (backtrack(i, j, 0, board, word)) return true;
            }
        }

        return false;
    }
};

// ─── Driver ──────────────────────────────────────────────────────────────────
int main() {
    Solution sol;

    // Test 1
    vector<vector<char>> b1 = {
        {'A', 'B', 'C', 'E'},
        {'S', 'F', 'C', 'S'},
        {'A', 'D', 'E', 'E'}
    };
    string w1 = "ABCCED";
    cout << "Test 1 — Result: " << (sol.exist(b1, w1) ? "true" : "false") << endl;
    // Expected: true

    // Test 2
    string w2 = "SEE";
    cout << "Test 2 — Result: " << (sol.exist(b1, w2) ? "true" : "false") << endl;
    // Expected: true

    // Test 3
    string w3 = "ABCB";
    cout << "Test 3 — Result: " << (sol.exist(b1, w3) ? "true" : "false") << endl;
    // Expected: false

    return 0;
}
